dblookupext/esdtSupply/corrections.go

```diff
@@ -0,0 +1,174 @@
+package esdtSupply
+
+import (
+	"bytes"
+	"errors"
+	"math/big"
+
+	"github.com/multiversx/mx-chain-go/config"
+	"github.com/multiversx/mx-chain-go/storage"
+)
+
+const (
+	supplyCorrectionPrefix = "supplyCorrection"
+)
+
+type supplyCorrectionProcessor struct {
+	shardID         uint32
+	logsProc        *logsProcessor
+	lateCorrections []config.SupplyCorrection
+}
+
+func newSupplyCorrectionProcessor(shardID uint32, logsProc *logsProcessor) *supplyCorrectionProcessor {
+	return &supplyCorrectionProcessor{
+		shardID:         shardID,
+		logsProc:        logsProc,
+		lateCorrections: make([]config.SupplyCorrection, 0),
+	}
+}
+
+func (scp *supplyCorrectionProcessor) applyLateCorrections(currentBlockNonce uint64) error {
+	if len(scp.lateCorrections) == 0 {
+		return nil
+	}
+
+	supplies := make(map[string]*SupplyESDT)
+	remainingCorrections := make([]config.SupplyCorrection, 0, len(scp.lateCorrections))
+	for _, supplyCorrection := range scp.lateCorrections {
+		if supplyCorrection.BlockNonce > currentBlockNonce {
+			remainingCorrections = append(remainingCorrections, supplyCorrection)
+			continue
+		}
+
+		err := scp.processSupplyCorrection(supplyCorrection, supplies)
+		if err != nil {
+			return err
+		}
+	}
+
+	scp.lateCorrections = remainingCorrections
+
+	err := scp.logsProc.saveSupplies(supplies)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (scp *supplyCorrectionProcessor) applySupplyCorrections(supplyCorrections []config.SupplyCorrection) error {
+	supplies := make(map[string]*SupplyESDT)
+	for _, supplyCorrection := range supplyCorrections {
+		shouldApply, err := scp.shouldApplyCorrectionAndSaveLateCorrections(supplyCorrection)
+		if err != nil {
+			return err
+		}
+		if !shouldApply {
+			continue
+		}
+
+		err = scp.processSupplyCorrection(supplyCorrection, supplies)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := scp.logsProc.saveSupplies(supplies)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (scp *supplyCorrectionProcessor) processSupplyCorrection(supplyCorrection config.SupplyCorrection, supplies map[string]*SupplyESDT) error {
+	tokenSupply, err := scp.logsProc.getESDTSupply([]byte(supplyCorrection.Token))
+	if err != nil {
+		return err
+	}
+
+	// if token supply was recomputed should not apply supply correction
+	if tokenSupply.RecomputedSupply {
+		return nil
```

Comment:
could you start before the supplyCorrectionBlock with the RecomputedSupply = true, the "bug" manifests after and when it is time to do the supplyCorrection, you think it was recompted?